lib/pgBackRest/Storage/Local.pm

```diff
@@ -0,0 +1,775 @@
+####################################################################################################################################
+# Local storage object that uses drivers for file system operations.
+####################################################################################################################################
+package pgBackRest::Storage::Local;
+use parent 'pgBackRest::Storage::Base';
+
+use strict;
+use warnings FATAL => qw(all);
+use Carp qw(confess);
+use English '-no_match_vars';
+
+use File::Basename qw(dirname);
+
+use pgBackRest::Common::Exception;
+use pgBackRest::Common::Log;
+use pgBackRest::Common::String;
+use pgBackRest::Storage::Filter::Sha;
+
+####################################################################################################################################
+# new
+####################################################################################################################################
+sub new
+{
+    my $class = shift;
+
+    # Assign function parameters, defaults, and log debug info
+    my
+    (
+        $strOperation,
+        $strPathBase,
+        $oDriver,
+        $hRule,
+        $bAllowTemp,
+        $strTempExtension,
+        $strDefaultPathMode,
+        $strDefaultFileMode,
+        $lBufferMax,
+    ) =
+        logDebugParam
+        (
+            __PACKAGE__ . '->new', \@_,
+            {name => 'strPathBase'},
+            {name => 'oDriver'},
+            {name => 'hRule', optional => true},
+            {name => 'bAllowTemp', optional => true, default => true},
+            {name => 'strTempExtension', optional => true, default => 'tmp'},
+            {name => 'strDefaultPathMode', optional => true, default => '0750'},
+            {name => 'strDefaultFileMode', optional => true, default => '0640'},
+            {name => 'lBufferMax', optional => true},
+        );
+
+    # Create class
+    my $self = $class->SUPER::new({lBufferMax => $lBufferMax});
+    bless $self, $class;
+
+    $self->{strPathBase} = $strPathBase;
+    $self->{oDriver} = $oDriver;
+    $self->{hRule} = $hRule;
+    $self->{bAllowTemp} = $bAllowTemp;
+    $self->{strTempExtension} = $strTempExtension;
+    $self->{strDefaultPathMode} = $strDefaultPathMode;
+    $self->{strDefaultFileMode} = $strDefaultFileMode;
+
+    # Set temp extension in driver
+    if (defined($self->{oDriver}))
```

Comment:
CSHANG Why is this check necessary? oDriver is a required parameter, if not defined it would blow up when the function is called, no? I don't see how we could test this.